File: BubbleSort.java
package com.proglang;

import java.util.Arrays;
import java.util.Scanner;

public class BubbleSort {
    public static void main(String[] args) {

        /*  Bubble Sort
            Time Complexity -
                Best Case - Omega(n)
                Average Case - Theta(n^2)
                Worst Case - BigO(n^2)
         */

        Scanner scanner = new Scanner(System.in);
        System.out.print("n: ");
        int n = scanner.nextInt();

        scanner.nextLine();

        String[] s = scanner.nextLine().split(" ");
        int[] array = new int[n];

        for(int i=0; i<n; i++) {
            array[i] = Integer.parseInt(s[i]);
        }

        int tmp, swapCount=0;
        for (int i=0; i<array.length; i++) {
            for (int k=0; k<array.length-i-1; k++) {
                if (array[k] > array[k+1]) {
                    tmp = array[k];
                    array[k] = array[k+1];
                    array[k+1] = tmp;
                    swapCount++;
                    System.out.println("swap " + swapCount + " : " + Arrays.toString(array));
                }
            }
        }

        System.out.println("Sorted Array : " + Arrays.toString(array));
        System.out.println("swapCount : " + swapCount);

        scanner.close();
    }
}

File: DSA - Python/Sorting/bubble_sort.py

# Bubble sort
# complexity - O(n^2) time | O(1) space | O(n^2) best time
def bubble_sort(arr):
    for i in range(len(arr)):
        for j in range(0, len(arr)-i-1):
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]
    return arr


# complexity - O(n^2) time | O(1) space | O(n) best time
def bubble_sort_best(array):
    is_sorted = False
    counter = 0
    while not is_sorted:
        is_sorted = True
        for i in range(0, len(array)-1-counter):
            if array[i] > array[i+1]:
                swap(i, i+1, array)
                is_sorted = False
        counter += 1
    return array


def swap(i, j, array):
    array[i], array[j] = array[j], array[i]
    # swap in diff. languages
    # tmp = array[i]
    # array[i] = array[j]
    # array[j] = tmp


arr = list(map(int, input().split(' ')))
print(bubble_sort(arr))
print(bubble_sort_best(arr))